### PR TITLE
id_set + dependencies UT 2nd stage

### PR DESCRIPTION
**Pull request by teizenman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Entities completed
- [x] Integration
- [ ] Script
- [ ] Playbooks
- [x] Mapper
- [x] Classifier
- [x] Incident Type
- [x] Indicator Type
- [x] Layout
- [ ] New Layout
- [x] Incident Field

## Related Issues
Related: https://github.com/demisto/etc/issues/25887

### CODE REVIEW

File: TestSuite/repo.py [hunk 1]
```diff
@@ -51,6 +51,75 @@ def __init__(self, tmpdir: Path):
         self.id_set = JSONBased(self._test_dir, 'id_set', '')
         self.id_set.write_json({})
 
+    def setup_one_pack(self, name):
```
Comment (ronykoz) [inline TestSuite/repo.py:54]: add docs about the things you create here

File: TestSuite/assets/default_playbook/playbook-sample_CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+## [Unreleased]
```
Comment (yaakovi) [inline TestSuite/assets/default_playbook/playbook-sample_CHANGELOG.md:1]: why do you need this file?

File: TestSuite/pack.py [hunk 3]
```diff
@@ -227,10 +249,91 @@ def create_indicator_field(
...
+            self,
+            name,
+            content: dict = None
+    ) -> JSONBased:
+        prefix = 'reputation'
+        indicator_type = self._create_json_based(name, prefix, content, dir_path=self._indicator_types)
+        self.indicator_types.append(indicator_type)
+        return indicator_type
+
+    def create_layout(
```
Comment (yaakovi) [inline TestSuite/pack.py:268]: we now also have `layoutscontainer`

File: TestSuite/playbook.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        self._repo = repo
+        self.repo_path = repo.path
+        self.is_test_playbook = is_test_playbook
+
+        self.path = str(tmpdir)
+        self.yml = YAML(tmpdir / f'{self.name}.yml', self._repo.path)
+
+        if not self.is_test_playbook:
+            self.readme = File(tmpdir / 'README.md', self._repo.path)
+            self.changelog = File(tmpdir / 'CHANGELOG.md', self._repo.path)
```
Comment (yaakovi) [inline TestSuite/playbook.py:23]: no need, we stop using those files

File: TestSuite/playbook.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        if yml is not None:
+            self.yml.write_dict(yml)
+        if not self.is_test_playbook and readme is not None:
+            self.readme.write(readme)
+        if not self.is_test_playbook and changelog is not None:
+            self.changelog.write(changelog)
+
+    def create_default_playbook(self):
+        default_playbook_dir = 'assets/default_playbook'
+        yml = open(suite_join_path(default_playbook_dir, 'playbook-sample.yml'))
```
Comment (yaakovi) [inline TestSuite/playbook.py:49]: why not use a `with` statement?

File: demisto_sdk/commands/create_id_set/tests/update_id_set_test.py [hunk 6]
```diff
@@ -428,30 +427,6 @@ def test_get_layout_data_no_incident_types_and_fields():
         assert 'incident_and_indicator_types' in result.keys()
         assert 'incident_and_indicator_fields' not in result.keys()
 
-    @staticmethod
```
Comment (yaakovi) [inline demisto_sdk/commands/create_id_set/tests/update_id_set_test.py:431]: why did you remove this?

File: TestSuite/assets/default_playbook/playbook-sample_CHANGELOG.md (hunk 2 above)
Comment (teizenman) [reply TestSuite/assets/default_playbook/playbook-sample_CHANGELOG.md:1]: Did that before the transition.

File: demisto_sdk/commands/create_id_set/tests/update_id_set_test.py (hunk 6 above)
Comment (teizenman) [reply demisto_sdk/commands/create_id_set/tests/update_id_set_test.py:431]: My mistake

**Comment by teizenman** [comment]
Not all entities are implemented yet, but I will do it in another PR so @yaakovi can complete his task

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 7]
```diff
@@ -111,18 +111,19 @@ def make_indicator_type_depend_on_script_enhancement(indicator_type: JSONBased,
...
-        integration.yml.update(
-            {
-                'script': {
-                    'commands': [{
-                        'name': 'ip'
-                    }]
-                }
-            }
-        )
+    # Optional dependency
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py:114]: why was that removed?

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 8]
```diff
@@ -182,16 +183,17 @@ def make_layout_depend_on_incident_indicator_field(layout: JSONBased, indicators
...
-    def make_incident_field_depend_on_incident_type_associated(incident_field: JSONBased, incident_type: JSONBased):
-        incident_type_id = incident_type.read_json_as_dict().get('id')
-        incident_field.update({'associatedTypes': [incident_type_id]})
-
-    @staticmethod
-    def make_incident_field_depend_on_incident_type_system_associated(incident_field: JSONBased,
-                                                                      incident_type: JSONBased):
-        incident_type_id = incident_type.read_json_as_dict().get('id')
-        incident_field.update({'systemAssociatedTypes': [incident_type_id]})
+    # Ignored by yaakovi
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py:186]: :rigshi-sade:

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 9]
```diff
@@ -298,11 +300,14 @@ def run_random_methods(repo, current_pack, current_methods_pool, number_of_metho
         all_dependencies = all_dependencies.union(dependencies)
         method(**args)
 
+        # print(dependencies)
+        # print(method.__name__)
+
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py]: remove those lines

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 10]
```diff
@@ -256,7 +258,7 @@ def create_inputs_for_method(repo, current_pack, inputs_arguments):
                      get_entity_by_pack_number_and_entity_type(repo, current_pack, inputs_arguments[0])}
 
     inputs_arguments = inputs_arguments[1:]
-    number_of_packs = len(repo.packs)
+    number_of_packs = len(repo.packs) - 1
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py:261]: why?

File: demisto_sdk/tests/integration_tests/find_dependencies_integration_test.py [hunk 11]
```diff
@@ -46,6 +46,8 @@ def test_integration_find_dependencies__sanity(mocker, repo):
 
     # Change working dir to repo
     with ChangeCWD(integration.repo_path):
+        import demisto_sdk.commands.common.update_id_set as uis
+        mocker.patch.object(uis, 'cpu_count', return_value=1)
```
Comment (yaakovi) [inline demisto_sdk/tests/integration_tests/find_dependencies_integration_test.py:50]: why?

File: demisto_sdk/commands/common/tests/dependencies_test.py (hunk 7 above)
Comment (teizenman) [reply demisto_sdk/commands/common/tests/dependencies_test.py:114]: At the beginning I thought that it wont be included since its an optional dependency.
Put it back

File: demisto_sdk/commands/common/tests/dependencies_test.py (hunk 10 above)
Comment (teizenman) [reply demisto_sdk/commands/common/tests/dependencies_test.py:261]: Since I add the `CommonTypes` pack, and I dont want any of its entities to be included in the regular dependency flow I used this line to ignore it.
I will add a comment

File: demisto_sdk/tests/integration_tests/find_dependencies_integration_test.py (hunk 11 above)
Comment (teizenman) [reply demisto_sdk/tests/integration_tests/find_dependencies_integration_test.py:50]: One of the reasons that Circle's froze on 3.7 is the high usage of processing power, so I reduced it.

`pool = Pool(processes=cpu_count() * 2)` is the line that in charge of the multiprocessing initiation, so changing `cpu_count` return value to 1 still gives you multiprocessing but with only 2 processors, and not the maximum amount.

Comment (yaakovi) [reply demisto_sdk/tests/integration_tests/find_dependencies_integration_test.py:50]: comment that next to the mock

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 12]
```diff
@@ -285,7 +286,7 @@ def run_random_methods(repo, current_pack, current_methods_pool, number_of_metho
     all_dependencies = set()
 
     for i in range(number_of_methods_to_choose):
-        chosen_method_index = random.sample(range(1, len(current_methods_pool)), 1)[0]
+        chosen_method_index = random.sample(range(len(current_methods_pool)), 1)[0]
         chosen_method = current_methods_pool[chosen_method_index]
         current_methods_pool.remove(chosen_method)
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py]: since you are choosing just one and you don't care about the index, you can use `random.choice`

File: demisto_sdk/commands/common/tests/dependencies_test.py [hunk 13]
```diff
@@ -283,10 +295,22 @@ def create_inputs_for_method(repo, current_pack, inputs_arguments):
...
+    Returns:
+        Set. All the packs' names that `current_pack` should depend on.
+    """
     all_dependencies = set()
 
     for i in range(number_of_methods_to_choose):
-        chosen_method_index = random.sample(range(len(current_methods_pool)), 1)[0]
+        chosen_method_index = random.choice(range(len(current_methods_pool)))
         chosen_method = current_methods_pool[chosen_method_index]
         current_methods_pool.remove(chosen_method)
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/dependencies_test.py]: i meant
```suggestion
        chosen_method = random.choice(current_methods_pool)
        current_methods_pool.remove(chosen_method)
```

Comment (teizenman) [reply demisto_sdk/commands/common/tests/dependencies_test.py]: That is smarter
